### ISSUE TITLE
Conversion of `Scalar` to `double`

### ISSUE THREAD

**Issue by yingjerkao** [kaihsin/Cytnx#155]
I am working to use the newer oneAPI Intel compiler (`icpx`) to compile cytnx instead of the classical Intel compilers (`icpc`) included in the HPC toolkit. 

I fixed some obvious errors due to the more strict llvm compiler checks but  ran into this and don't know where the problem comes from. It's related to the conversion of the `Scalar` type to `double`.

`Out` is a `Tensor` object

```
/home/yjkao/Cytnx/src/Physics.cpp:33:50: error: no viable overloaded '='
            Out.at<cytnx_complex128>({a, a - 1}) =
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/Physics.cpp:36:50: error: no viable overloaded '='
            Out.at<cytnx_complex128>({a, a + 1}) =
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
```

**Comment by jeffry1829** [comment]
Hi, Prof Kao
May I know how do you resolve this?
I thought we need to fix type conversion problems everywhere in this project

**Comment by yingjerkao** [comment]
I did a static_cast<double> to force the conversion. It is related to the complex addition, but I am not sure exactly why. The same issue occurs in `Add_internal.cpp`

**Comment by jeffry1829** [comment]
I thought we need to explicitly conversion in dozens of files
So it's only 2 files?

**Comment by yingjerkao** [comment]
I pushed a branch `oneapi` to address some of the issues. The latest is 5d12cfbdf1eae3b23c976fb6e223faf23e737a5c. 

I find it odd that it seems to occur at the complex addition overload.

**Comment by yingjerkao** [comment]
Also, `Tensor` indices allows for negative indices but people are used to using `uint` as index. This causes narrowing errors in compiling the initialization lists. Also, `Tesnor.size()` returns `unit` but it is commonly used for indexing, thus the similar errors. 

These are minor design details and can be turned off by compiler flags. But I think it would be nice to avoid these in the base code.

**Comment by yingjerkao** [comment]
`/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:88:19: error: no viable overloaded '='
          _out[i] = _Lin[0] + _Rin[i];
          ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:95:19: error: no viable overloaded '='
          _out[i] = _Lin[i] + _Rin[0];
          ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:103:21: error: no viable overloaded '='
            _out[i] = _Lin[i] + _Rin[i];
            ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:127:21: error: no viable overloaded '='
            _out[i] = _Lin[cartesian2c(vec_map(tmpv, invmapper_L), old_accu_shapeL)] +
            ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:208:19: error: no viable overloaded '='
          _out[i] = _Lin[0] + _Rin[i];
          ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:215:19: error: no viable overloaded '='
          _out[i] = _Lin[i] + _Rin[0];
          ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:223:21: error: no viable overloaded '='
            _out[i] = _Lin[i] + _Rin[i];
            ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:247:21: error: no viable overloaded '='
            _out[i] = _Lin[cartesian2c(vec_map(tmpv, invmapper_L), old_accu_shapeL)] +
            ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:268:19: error: no viable overloaded '='
          _out[i] = _Lin[0] + _Rin[i];
          ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:275:19: error: no viable overloaded '='
          _out[i] = _Lin[i] + _Rin[0];
          ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:283:21: error: no viable overloaded '='
            _out[i] = _Lin[i] + _Rin[i];
            ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:307:21: error: no viable overloaded '='
            _out[i] = _Lin[cartesian2c(vec_map(tmpv, invmapper_L), old_accu_shapeL)] +
            ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:328:19: error: no viable overloaded '='
          _out[i] = _Lin[0] + _Rin[i];
          ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:335:19: error: no viable overloaded '='
          _out[i] = _Lin[i] + _Rin[0];
          ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:343:21: error: no viable overloaded '='
            _out[i] = _Lin[i] + _Rin[i];
            ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:367:21: error: no viable overloaded '='
            _out[i] = _Lin[cartesian2c(vec_map(tmpv, invmapper_L), old_accu_shapeL)] +
            ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:388:19: error: no viable overloaded '='
          _out[i] = _Lin[0] + _Rin[i];
          ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:395:19: error: no viable overloaded '='
          _out[i] = _Lin[i] + _Rin[0];
          ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
/home/yjkao/Cytnx/src/linalg/linalg_internal_cpu/Add_internal.cpp:403:21: error: no viable overloaded '='
            _out[i] = _Lin[i] + _Rin[i];
            ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1285:7: note: candidate function not viable: no known conversion from 'Scalar' to 'double' for 1st argument
      operator=(double __d)
      ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1321:37: note: candidate function not viable: no known conversion from 'Scalar' to 'const complex<double>' for 1st argument
      _GLIBCXX14_CONSTEXPR complex& operator=(const complex&) = default;
                                    ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/complex:1326:9: note: candidate template ignored: could not match 'complex<_Tp>' against 'Scalar'
        operator=(const complex<_Tp>& __z)
        ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
`

**Comment by yingjerkao** [comment]
@kaihsin  I wonder if this `Scalar` is related to the `Scalar` type defined in `Storage.hpp`?

**Comment by yingjerkao** [comment]
I think this bug is related to this issue
https://github.com/pybind/pybind11/issues/1818

In short,  llvm/clang does not, by default, define __cpp_sized_deallocation which libstdc++ uses. GCC does define this, and one need to force clang to with `-fsized-deallocation`.

**Comment by yingjerkao** [comment]
This commit fixes the issue a250c5566172ce8799765c22c2fcf2506b9189d4 
Also, I remove the USE_ICPC flag for detecting Intel compilers. Users are recommended to use CMAKE_C_COMPILER and CMAKE_CXX_COMPILER to assign non-default  compilers.
